google-cloud-firestore/src/test/java/com/google/cloud/firestore/RecursiveDeleteTest.java

```diff
@@ -0,0 +1,678 @@
+    responseStubber.verifyAllRequestsSent();
+  };
+
+  @Test
+  public void handlesMultipleCallsToRecursiveDelete() throws Exception {
+    final List<Answer<RunQueryResponse>> runQueryResponses =
+        new ArrayList<>(
+            Arrays.asList(
+                queryResponse(fullDocumentPath("a")),
+                queryResponse(fullDocumentPath("b")),
+                queryResponse(fullDocumentPath("c"))));
+
+    doAnswer(runQueryResponses.get(0))
+        .doAnswer(runQueryResponses.get(1))
+        .doAnswer(runQueryResponses.get(2))
+        .when(firestoreMock)
+        .streamRequest(
+            runQueryCapture.capture(),
+            streamObserverCapture.capture(),
+            Matchers.<ServerStreamingCallable>any());
+
+    ResponseStubber responseStubber =
+        new ResponseStubber() {
+          {
+            put(batchWrite(delete("coll/a")), successResponse(1));
+            put(batchWrite(delete("coll/b")), successResponse(2));
+            put(batchWrite(delete("coll/c")), successResponse(3));
+          }
+        };
+    responseStubber.initializeStub(batchWriteCapture, firestoreMock);
+
+    firestoreMock.recursiveDelete(firestoreMock.collection("a")).get();
+    firestoreMock.recursiveDelete(firestoreMock.collection("b")).get();
+    firestoreMock.recursiveDelete(firestoreMock.collection("c")).get();
+    responseStubber.verifyAllRequestsSent();
+  }
+
+  @Test
+  public void usesSameBulkWriterInstanceAcrossCalls() throws Exception {
+    doAnswer(emptyQueryResponse())
+        .when(firestoreMock)
+        .streamRequest(
+            runQueryCapture.capture(),
+            streamObserverCapture.capture(),
+            Matchers.<ServerStreamingCallable>any());
+
+    final int[] callCount = {0};
+    final BulkWriter bulkWriter = firestoreMock.bulkWriter();
+    doAnswer(
+            new Answer<BulkWriter>() {
+              public BulkWriter answer(InvocationOnMock mock) throws Throwable {
+                callCount[0]++;
+                return bulkWriter;
+              }
+            })
+        .when(firestoreMock)
+        .bulkWriter();
+
+    firestoreMock.recursiveDelete(firestoreMock.collection("foo")).get();
+    firestoreMock.recursiveDelete(firestoreMock.collection("boo")).get();
+    firestoreMock.recursiveDelete(firestoreMock.collection("moo")).get();
+
+    // Only the first recursiveDelete() call should have called the constructor. Subsequent calls
+    // should have used the same bulkWriter.
+    assertEquals(1, callCount[0]);
+  }
+
+  @Test
+  public void throwsErrorIfBulkWriterInstanceIsClosed() throws Exception {
+    bulkWriter = firestoreMock.bulkWriter();
+    bulkWriter.close();
+    try {
+      firestoreMock.recursiveDelete(firestoreMock.collection("foo"), bulkWriter);
+      fail("Operation should have failed in test");
+    } catch (Exception e) {
+      assertTrue(e.getMessage().contains("BulkWriter has already been closed"));
+    }
+  }
+}
```

Comment:
I know it will be pretty tough to test but we do not have a test that tests that we send the second request when we pass the MIN_PENDING_OPS threshold.